Upgrading to Bootstrap 5 and removing dependency on JQuery.

sergiotarxz opened the pull request:
### Summary
Migrating to Boostrap 5 and removing dependency on JQuery.

### Motivation
Issue #1975

### References
#1975

sergiotarxz commented:
![Screenshot from 2024-02-14 00-51-58](https://github.com/mojolicious/mojo/assets/25019174/549a64ea-e04b-4dba-b1c9-42d3e5ab8b7e)
![Screenshot from 2024-02-14 00-51-50](https://github.com/mojolicious/mojo/assets/25019174/625fa53b-6678-46ae-ba7a-49e7d0f1ee68)
![Screenshot from 2024-02-14 00-51-32](https://github.com/mojolicious/mojo/assets/25019174/c1ffe787-0fcc-49e8-9aa7-216c6b054dc7)
![Screenshot from 2024-02-14 00-51-16](https://github.com/mojolicious/mojo/assets/25019174/c11eefc7-f769-47ef-84d5-f9a65ba55a9e)
![Screenshot from 2024-02-14 00-50-54](https://github.com/mojolicious/mojo/assets/25019174/dbda8195-7b0b-4364-978c-bfe27e5350bb)
![Screenshot from 2024-02-14 00-50-44](https://github.com/mojolicious/mojo/assets/25019174/cd4e7142-d19d-4d77-ae38-7dc9f47c825f)
![Screenshot from 2024-02-14 00-50-33](https://github.com/mojolicious/mojo/assets/25019174/62eeacf2-9d8d-40c7-9550-623a20263190)
![Screenshot from 2024-02-14 00-50-21](https://github.com/mojolicious/mojo/assets/25019174/a654689d-2565-4b15-b7c6-9f565672f841)
Right new Left Old.

kraih commented on lib/Mojolicious/resources/public/mojo/mojo.css:
```diff
@@ -1,10 +1,17 @@
   height:100%;
 }
+body {
+  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
```
That line looks inconsistent.

kraih commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -100,18 +100,25 @@
-          const text = $(handle + ' div.tap').text();
+        const handleEl = document.querySelector(handle);
+        if (handleEl === null || drawer === null) {
+            console.error(`${handle} or ${drawer} are null`);
```
Why do we care?

kraih commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -170,24 +177,36 @@
-                      });
+                      const showcase = document.querySelector('#showcase');
+                      if (showcase === null) {
+                        console.error('#showcase is null');
```
Why do we care about this?

kraih replied:
You're overthinking this. It will just be dead code.

sergiotarxz replied:
Said and done, fixing it.

kraih replied:
I don't want to see `console.*` calls unless they actually provide information that an exception would not. And not for things that can't really happen.

kraih replied:
Here for `#showcase` not to exist someone else would have to have really messed up with another patch. Preparing for that just doesn't make sense.

sergiotarxz replied:
Removed, sorry for the incovenience.

sergiotarxz commented:
Also this pull requests fixes a bug where `tap for more` was not changing to `tap for less` on click.

sergiotarxz reviewed (commented):
My comments

sergiotarxz commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -17,9 +16,10 @@
-        $('[data-toggle="tooltip"]').tooltip()
-      });
+      const tooltipElements = document.querySelectorAll('[data-toggle="tooltip"]');
+      for (const tooltipElement of tooltipElements) {
```
There is only a tooltip, I will remove the document.querySelectorAll, it was defensive programming too.

kraih commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -100,18 +98,22 @@
-          $(handle + ' div.tap').text(text);
+          const divTap = document.querySelector(handle + ' div.tap');
+          if (divTap !== null) {
+            divTap.innerText = text;
```
I like to format these as:
```js
if (divTap !== null) divTap.innerText = text;
```
Instead of blocks.

sergiotarxz replied:
Done.

sergiotarxz commented on t/mojolicious/rebased_lite_app.t:
```diff
@@ -104,7 +102,6 @@ done_testing();
 __DATA__
 @@ root.html.ep
 %= $c->req->url->base
-%= javascript '/mojo/jquery/jquery.js'
```
Do you want me to preserve the file jquery.js in the repository? Or only to preserve this concrete test case? May I replace it with the bootstrap javascript?

kraih replied:
Only the test case.

kraih replied:
The file is never even requested.

sergiotarxz replied:
Restoring the file to its original contents.

sergiotarxz commented on t/mojolicious/exception_lite_app.t:
```diff
@@ -336,7 +336,6 @@ subtest 'Text exceptions' => sub {
 
 subtest 'Bundled static files' => sub {
-  $t->get_ok('/mojo/jquery/jquery.js')->status_is(200)->content_type_is('application/javascript');
 
```
Done

kraih commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -170,24 +170,30 @@
+                      const insightEl = document.querySelector('#insight');
+                      insightEl.classList.toggle('d-none');
+                    }
+                    toggleInsight();
```
Why is this a function?

sergiotarxz replied:
Because I was returning if I had an error, it is no longer a function now, sorry.

sergiotarxz commented:
There is a screenshot I have not posted
![image](https://github.com/mojolicious/mojo/assets/25019174/484deaad-1d8e-4b3d-9814-f110fc38bf73)

kraih commented on lib/Mojolicious/resources/templates/mojo/debug.html.ep:
```diff
@@ -100,18 +98,20 @@
-          const text = $(handle + ' div.tap').text();
+        const handleEl = document.querySelector(handle);
+        handleEl.addEventListener('click', () => {
+          drawerEl = document.querySelector(drawer);
```
Where does `drawerEl` come from? I'm not seeing its scope.

sergiotarxz replied:
It was an error, I added 'use strict'; to the javascript to prevent this kind of error.

sergiotarxz commented:
I added use strict to the javascript and discovered a bunch of errors.

sergiotarxz commented:
I just discovered there was a code path I was not testing, I tested it and fixed the following:
* Added popper.js. (Required by tooltip)
* Ensured tooltips actually work.
* Migrated a bunch of data- to data-bs

sergiotarxz commented:
@kraih Would you be in favor of moving the embedded javascript in the template to a file? Maybe in another pr.
